core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/common/AbstractUserStoreManager.java

```diff
@@ -10180,6 +10214,34 @@ private AuthenticationResult generateUserStoreChainWithID(String preferredUserNa
             String preferredUserNameValue, Object credential, String profileName, boolean domainProvided,
             List<String> userStorePreferenceOrder) throws UserStoreException {
 
+        // If domain name is provided, directly authenticate using the corresponding user store.
+        if (domainProvided) {
+            String providedDomainName = this.getUserStore(preferredUserNameValue).getDomainName();
+            // Check whether provided domain is in the preference list.
+            if (!userStorePreferenceOrder.contains(providedDomainName)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Authentication failure. Invalid username or password is provided.");
+                }
+                handleOnAuthenticateFailure(ErrorMessages.ERROR_CODE_ERROR_WHILE_AUTHENTICATION.getCode(),
+                        String.format(ErrorMessages.ERROR_CODE_ERROR_WHILE_AUTHENTICATION.getMessage(),
+                                "Authentication failed"), preferredUserNameValue, credential);
+                throw new UserStoreException("Authentication failed. Invalid username or password.");
+            }
+            UserStoreManager userStoreManager = this.getSecondaryUserStoreManager(providedDomainName);
+            if (!(userStoreManager instanceof AbstractUserStoreManager)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("UserStoreManager is not an instance of AbstractUserStoreManager. Hence " +
+                            "authenticate the user through all the available user store list.");
+                }
+                return authenticateInternalWithID(preferredUserNameClaim, preferredUserNameValue, credential,
+                        profileName, domainProvided);
+            }
+            IterativeUserStoreManager iterativeUserStoreManager = new IterativeUserStoreManager(
+                    (AbstractUserStoreManager) userStoreManager);
+            return iterativeUserStoreManager.
+                    authenticateWithID(preferredUserNameClaim, preferredUserNameValue, credential, profileName);
+        }
```

Comment:
Is it possible to extract this logic to a method and reuse?

Reply:
Addressed in https://github.com/wso2/carbon-kernel/pull/2765